### DISCUSSION TITLE
The true power of sealed classes – Blog

### DISCUSSION THREAD

**Discussion by giscus** [jiripudil/jiripudil.cz#103] [Blog discussions]
# The true power of sealed classes – Blog

A full-stack web developer from Brno

https://jiripudil.cz/blog/true-power-of-sealed-classes

**Comment by elieux** [comment]
Hello. Some of this looks like a work around for a bad choice of language primitive though I'm open to arguments.

The idea of interfaces being limited in their implementations sounds very foreign to me. I've never been convinced even with the "final class" thing. The example here would be better served at least by tagged unions, as you mentioned, but polymorphism would be even better. The match+instanceof pattern definitely looks like a (unnecessary?) replacement for polymorphism. One argument against that I can imagine is dividing responsibility, but that could still have a solution.

One other concept that kept jumping into mind when reading was "internal", e.g. interfaces (or anything else) whose usage would be disallowed outside the package/library.

**Reply by jiripudil** [reply]
Hi, sorry for the late reply, I've had it on my todo list, but then life kept happening 😬

> The example here would be better served at least by tagged unions

Sealed classes _are_ tagged unions, essentially, perhaps only seen from a different perspective. You could also think of the sealed interface as a type alias of the union of the permitted implementations, just like in TypeScript's discriminated union types. These are all representations of the same concept.

Unfortunately, PHP doesn't have native tagged unions with pattern matching or type aliases (yet 🤞). Having a common, sealed interface looks like the closest practical thing we have at our disposal for the time being. I admit the match+instanceof is not particularly good-looking, but that's what we have 🤷

> but polymorphism would be even better (...) One argument against that I can imagine is dividing responsibility, but that could still have a solution.

The main one, honestly. Lots of domain use cases around the sealed interface in this post can benefit from polymorphism, starting with state transitions. But I don't think a domain class should be concerned about how it is represented in a GUI or how it is serialized within an API response. Hence a function that is entirely external to the domain layer. I'm genuinely curious about what solution you would suggest!

> One other concept that kept jumping into mind when reading was "internal", e.g. interfaces (or anything else) whose usage would be disallowed outside the package/library.

Absolutely. But the use case is different, I think. Internal classes and interfaces are often meant to be _entirely_ hidden from the rest of the application. Sealed classes and interfaces merely restrict who can extend or implement them, while still allowing developers to use them in a type hint in the application code.
